File: radar-commons-android/src/main/java/org/radarcns/android/util/BatteryLevelReceiver.java
/*
 * Copyright 2017 The Hyve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarcns.android.util;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.support.annotation.NonNull;

import java.util.Objects;

import static android.content.Intent.ACTION_BATTERY_CHANGED;
import static android.os.BatteryManager.EXTRA_LEVEL;
import static android.os.BatteryManager.EXTRA_PLUGGED;
import static android.os.BatteryManager.EXTRA_SCALE;

/** Keep track of battery level events. */
public class BatteryLevelReceiver extends BroadcastReceiver implements SpecificReceiver {
    protected final Context context;
    private final BatteryLevelListener listener;
    private float level;
    private boolean isPlugged;

    public BatteryLevelReceiver(@NonNull Context context, BatteryLevelListener listener) {
        this.context = Objects.requireNonNull(context);
        this.listener = listener;
        this.level = 1.0f;
        this.isPlugged = true;
    }

    /** Latest battery level in range [0, 1]. */
    public float getLevel() {
        return level;
    }

    /** Latest value whether the device is plugged into a charger. */
    public boolean isPlugged() {
        return isPlugged;
    }

    /** Returns true if the current battery level is at least the given level, or that the device is
     * currently plugged in. Returns false otherwise. */
    public boolean hasMinimumLevel(float minLevel) {
        return level >= minLevel || isPlugged;
    }

    public void onReceive(Context context, Intent intent) {
        if (intent == null || !Objects.equals(intent.getAction(), ACTION_BATTERY_CHANGED)) {
            return;
        }
        int level1 = intent.getIntExtra(EXTRA_LEVEL, -1);
        int scale = intent.getIntExtra(EXTRA_SCALE, -1);

        this.level = level1 / (float)scale;
        this.isPlugged = intent.getIntExtra(EXTRA_PLUGGED, 0) > 0;

        if (listener != null) {
            listener.onBatteryLevelChanged(this.level, this.isPlugged);
        }
    }

    public void register() {
        Intent init = context.registerReceiver(this, new IntentFilter(ACTION_BATTERY_CHANGED));
        onReceive(context, init);
    }

    public void unregister() {
        context.unregisterReceiver(this);
    }

    public interface BatteryLevelListener {
        /**
         * Latest battery level reported by the system.
         * @param level battery level in range [0, 1]
         * @param isPlugged whether the device is plugged into a charger
         */
        void onBatteryLevelChanged(float level, boolean isPlugged);
    }
}

File: radar-commons-android/src/main/java/org/radarbase/android/util/BatteryLevelReceiver.kt
/*
 * Copyright 2017 The Hyve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarbase.android.util

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.Intent.ACTION_BATTERY_CHANGED
import android.content.IntentFilter
import android.os.BatteryManager.*

/** Keep track of battery level events.  */
class BatteryLevelReceiver(
        private val context: Context,
        var listener: ((level: Float, isPlugged: Boolean) -> Unit)? = null) : SpecificReceiver {
    /** Latest battery level in range [0, 1].  */
    var level: Float = 1.0f
        private set
    /** Latest value whether the device is plugged into a charger.  */
    var isPlugged: Boolean = true
        private set

    constructor(context: Context, listener: BatteryLevelListener) : this(context, listener::onBatteryLevelChanged)

    private var receiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent?) {
            if (intent?.action == ACTION_BATTERY_CHANGED) {
                val level1 = intent.getIntExtra(EXTRA_LEVEL, -1)
                val scale = intent.getIntExtra(EXTRA_SCALE, -1)

                level = level1 / scale.toFloat()
                isPlugged = intent.getIntExtra(EXTRA_PLUGGED, 0) > 0

                notifyListener()
            }
        }
    }

    /** Returns true if the current battery level is at least the given level, or that the device is
     * currently plugged in. Returns false otherwise.  */
    fun hasMinimumLevel(minLevel: Float): Boolean = level >= minLevel || isPlugged

    override fun register() {
        val init = context.registerReceiver(receiver, IntentFilter(ACTION_BATTERY_CHANGED))
        receiver.onReceive(context, init)
    }

    override fun notifyListener() {
        listener?.let { it(level, isPlugged) }
    }

    override fun unregister() {
        context.unregisterReceiver(receiver)
    }

    interface BatteryLevelListener {
        /**
         * Latest battery level reported by the system.
         * @param level battery level in range [0, 1]
         * @param isPlugged whether the device is plugged into a charger
         */
        fun onBatteryLevelChanged(level: Float, isPlugged: Boolean)
    }
}
